### ISSUE TITLE
Incompatible with Charm

### ISSUE THREAD

**Issue by InsaanNotFound** [YTG1234/fabric-recipe-conditions#2]
Recipe Conditions does the same thing again, but with every single version of Charm mod >2.1.6

```
[12:42:08] [main/WARN]: Failed to load datapacks, can't proceed with server load
java.util.concurrent.ExecutionException: com.google.gson.JsonParseException: Failed to load recipe charm:atlas/atlas!
	at Not Enough Crashes deobfuscated stack trace.(1.16.4+build.7) ~[?:?]
	at java.util.concurrent.CompletableFuture.reportGet(Unknown Source) ~[?:1.8.0_271]
	at java.util.concurrent.CompletableFuture.get(Unknown Source) ~[?:1.8.0_271]
	at net.minecraft.client.MinecraftClient.method_29604(MinecraftClient:1959) ~[?:?]
	at net.minecraft.client.MinecraftClient.startIntegratedServer(MinecraftClient:1808) ~[?:?]
	at net.minecraft.client.MinecraftClient.startIntegratedServer(MinecraftClient:1748) ~[?:?]
	at net.minecraft.client.gui.screen.TitleScreen.start(TitleScreen:5655) ~[?:?]
	at net.minecraft.client.gui.screen.TitleScreen.mdccb8e4$lambda$drawMenuButton$3$3(TitleScreen:5619) ~[?:?]
	at net.minecraft.client.gui.widget.ButtonWidget.onPress(ButtonWidget:33) ~[?:?]
	at net.minecraft.client.gui.widget.AbstractPressableButtonWidget.onClick(AbstractPressableButtonWidget:16) ~[?:?]
	at net.minecraft.client.gui.widget.AbstractButtonWidget.mouseClicked(AbstractButtonWidget:149) ~[?:?]
	at net.minecraft.client.gui.ParentElement.mouseClicked(ParentElement:27) ~[?:?]
	at net.minecraft.client.gui.screen.TitleScreen.mouseClicked(TitleScreen:296) ~[?:?]
	at net.minecraft.client.Mouse.method_1611(Mouse:92) ~[?:?]
	at net.minecraft.client.gui.screen.Screen.wrapScreenError(Screen:435) ~[?:?]
	at net.minecraft.client.Mouse.onMouseButton(Mouse:92) ~[?:?]
	at net.minecraft.client.Mouse.method_22686(Mouse:162) ~[?:?]
	at net.minecraft.util.thread.ThreadExecutor.execute(ThreadExecutor:94) ~[?:?]
	at net.minecraft.client.Mouse.method_22684(Mouse:162) ~[?:?]
	at org.lwjgl.glfw.GLFWMouseButtonCallbackI.callback(GLFWMouseButtonCallbackI.java:36) ~[lwjgl-glfw-3.2.2.jar:build 10]
	at org.lwjgl.system.JNI.invokeV(Native Method) ~[lwjgl-3.2.2.jar:build 10]
	at org.lwjgl.glfw.GLFW.glfwPollEvents(GLFW.java:3101) ~[lwjgl-glfw-3.2.2.jar:build 10]
	at com.mojang.blaze3d.systems.RenderSystem.flipFrame(RenderSystem:102) ~[intermediary-minecraft-1.16.4-client.jar:?]
	at net.minecraft.client.util.Window.swapBuffers(Window:308) ~[?:?]
	at net.minecraft.client.MinecraftClient.render(MinecraftClient:1068) ~[?:?]
	at net.minecraft.client.MinecraftClient.run(MinecraftClient:681) ~[?:?]
	at net.minecraft.client.main.Main.main(Main:215) [intermediary-minecraft-1.16.4-client.jar:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_271]
	at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source) ~[?:1.8.0_271]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source) ~[?:1.8.0_271]
	at java.lang.reflect.Method.invoke(Unknown Source) ~[?:1.8.0_271]
	at net.fabricmc.loader.game.MinecraftGameProvider.launch(MinecraftGameProvider.java:224) [fabric-loader-0.10.8.jar:?]
	at net.fabricmc.loader.launch.knot.Knot.init(Knot.java:141) [fabric-loader-0.10.8.jar:?]
	at net.fabricmc.loader.launch.knot.KnotClient.main(KnotClient.java:27) [fabric-loader-0.10.8.jar:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.8.0_271]
	at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source) ~[?:1.8.0_271]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source) ~[?:1.8.0_271]
	at java.lang.reflect.Method.invoke(Unknown Source) ~[?:1.8.0_271]
	at org.multimc.onesix.OneSixLauncher.launchWithMainClass(OneSixLauncher.java:196) [NewLaunch.jar:?]
	at org.multimc.onesix.OneSixLauncher.launch(OneSixLauncher.java:231) [NewLaunch.jar:?]
	at org.multimc.EntryPoint.listen(EntryPoint.java:143) [NewLaunch.jar:?]
	at org.multimc.EntryPoint.main(EntryPoint.java:34) [NewLaunch.jar:?]
Caused by: com.google.gson.JsonParseException: Failed to load recipe charm:atlas/atlas!
	at net.minecraft.recipe.RecipeManager.handler$dok000$processConditions(RecipeManager:4548) ~[?:?]
	at net.minecraft.recipe.RecipeManager.apply(RecipeManager) ~[?:?]
	at net.minecraft.recipe.RecipeManager.apply(RecipeManager:35) ~[?:?]
	at net.minecraft.resource.SinglePreparationResourceReloadListener.method_18790(SinglePreparationResourceReloadListener:13) ~[?:?]
	at java.util.concurrent.CompletableFuture.uniAccept(Unknown Source) ~[?:1.8.0_271]
	at java.util.concurrent.CompletableFuture$UniAccept.tryFire(Unknown Source) ~[?:1.8.0_271]
	at java.util.concurrent.CompletableFuture$Completion.run(Unknown Source) ~[?:1.8.0_271]
	at net.minecraft.resource.ResourceReloader.method_18365(ResourceReloader:71) ~[?:?]
	at net.minecraft.util.thread.ThreadExecutor.executeTask(ThreadExecutor:144) ~[?:?]
	at net.minecraft.util.thread.ReentrantThreadExecutor.executeTask(ReentrantThreadExecutor:23) ~[?:?]
	at net.minecraft.util.thread.ThreadExecutor.runTask(ThreadExecutor:118) ~[?:?]
	at net.minecraft.util.thread.ThreadExecutor.runTasks(ThreadExecutor:127) ~[?:?]
	at net.minecraft.client.MinecraftClient.method_29604(MinecraftClient:1958) ~[?:?]
	... 38 more
Caused by: com.google.gson.JsonParseException: java.lang.IllegalArgumentException: Unknown condition minecraft:type!
	at io.github.ytg1234.recipeconditions.api.condition.SingleCondition.fromJson(SingleCondition.java:91) ~[fabric-recipe-conditions-0.3.1.jar:?]
	at io.github.ytg1234.recipeconditions.api.condition.EveryCondition.fromJson(EveryCondition.java:35) ~[fabric-recipe-conditions-0.3.1.jar:?]
	at io.github.ytg1234.recipeconditions.api.condition.AnyCondition.fromJson(AnyCondition.java:40) ~[fabric-recipe-conditions-0.3.1.jar:?]
	at net.minecraft.recipe.RecipeManager.handler$dok000$processConditions(RecipeManager:4542) ~[?:?]
	at net.minecraft.recipe.RecipeManager.apply(RecipeManager) ~[?:?]
	at net.minecraft.recipe.RecipeManager.apply(RecipeManager:35) ~[?:?]
	at net.minecraft.resource.SinglePreparationResourceReloadListener.method_18790(SinglePreparationResourceReloadListener:13) ~[?:?]
	at java.util.concurrent.CompletableFuture.uniAccept(Unknown Source) ~[?:1.8.0_271]
	at java.util.concurrent.CompletableFuture$UniAccept.tryFire(Unknown Source) ~[?:1.8.0_271]
	at java.util.concurrent.CompletableFuture$Completion.run(Unknown Source) ~[?:1.8.0_271]
	at net.minecraft.resource.ResourceReloader.method_18365(ResourceReloader:71) ~[?:?]
	at net.minecraft.util.thread.ThreadExecutor.executeTask(ThreadExecutor:144) ~[?:?]
	at net.minecraft.util.thread.ReentrantThreadExecutor.executeTask(ReentrantThreadExecutor:23) ~[?:?]
	at net.minecraft.util.thread.ThreadExecutor.runTask(ThreadExecutor:118) ~[?:?]
	at net.minecraft.util.thread.ThreadExecutor.runTasks(ThreadExecutor:127) ~[?:?]
	at net.minecraft.client.MinecraftClient.method_29604(MinecraftClient:1958) ~[?:?]
	... 38 more
Caused by: java.lang.IllegalArgumentException: Unknown condition minecraft:type!
	at io.github.ytg1234.recipeconditions.api.condition.SingleCondition.fromJson(SingleCondition.java:91) ~[fabric-recipe-conditions-0.3.1.jar:?]
	at io.github.ytg1234.recipeconditions.api.condition.EveryCondition.fromJson(EveryCondition.java:35) ~[fabric-recipe-conditions-0.3.1.jar:?]
	at io.github.ytg1234.recipeconditions.api.condition.AnyCondition.fromJson(AnyCondition.java:40) ~[fabric-recipe-conditions-0.3.1.jar:?]
	at net.minecraft.recipe.RecipeManager.handler$dok000$processConditions(RecipeManager:4542) ~[?:?]
	at net.minecraft.recipe.RecipeManager.apply(RecipeManager) ~[?:?]
	at net.minecraft.recipe.RecipeManager.apply(RecipeManager:35) ~[?:?]
	at net.minecraft.resource.SinglePreparationResourceReloadListener.method_18790(SinglePreparationResourceReloadListener:13) ~[?:?]
	at java.util.concurrent.CompletableFuture.uniAccept(Unknown Source) ~[?:1.8.0_271]
	at java.util.concurrent.CompletableFuture$UniAccept.tryFire(Unknown Source) ~[?:1.8.0_271]
	at java.util.concurrent.CompletableFuture$Completion.run(Unknown Source) ~[?:1.8.0_271]
	at net.minecraft.resource.ResourceReloader.method_18365(ResourceReloader:71) ~[?:?]
	at net.minecraft.util.thread.ThreadExecutor.executeTask(ThreadExecutor:144) ~[?:?]
	at net.minecraft.util.thread.ReentrantThreadExecutor.executeTask(ReentrantThreadExecutor:23) ~[?:?]
	at net.minecraft.util.thread.ThreadExecutor.runTask(ThreadExecutor:118) ~[?:?]
	at net.minecraft.util.thread.ThreadExecutor.runTasks(ThreadExecutor:127) ~[?:?]
	at net.minecraft.client.MinecraftClient.method_29604(MinecraftClient:1958) ~[?:?]
	... 38 more
```
Error when game is trying to parse Atlas item crafting recipe. Here is text of Atlas crafting recipe, if this would help in any way.

```
{
  "type": "minecraft:crafting_shapeless",
  "ingredients": [
    {
      "tag": "charm:chests/normal"
    },
    {
      "item": "minecraft:book"
    },
    {
      "item": "minecraft:map"
    }
  ],
  "result": {
    "item": "charm:atlas",
    "count": 1
  },
  "conditions": [
    {
      "type": "charm:module_enabled",
      "module": "charm:atlas"
    }
  ]
}
```

**Comment by YTG1234** [comment]
@svenhjol might want to take a look at this

**Comment by YTG1234** [comment]
consider using my library instead of your own condition implementation

**Comment by YTG1234** [comment]
[The Charm conditions](https://github.com/svenhjol/Charm/blob/master/src/main/java/svenhjol/charm/mixin/RecipeManagerMixin.java)

**Comment by YTG1234** [comment]
Assuming Sven is reading this, here's how you can use my library:
- Use `Identifier`s for modules instead of `String`s (`String`s still work but if you specify it in the `Identifier` format in the recipe, why not in the `Map`).
- If you do that, a recipe condition could look like this:
```java
Registry.register(RecipeConds.RECIPE_CONDITION, new Identifier("charm", "module_enabled"),
    RecipeCondsUtil.stringParam(module -> {
        Map<Identifier, CharmModule> modules = ModuleHandler.getLoadedModules();
        return modules.get(new Identifier(module)).enabled;
    });
);
```
- And recipes would look like this:
```json5
{
// ...
  "conditions": [
    {
      "charm:module_enabled": "charm:module_name"
    }
  ]
}
```

**Comment by YTG1234** [comment]
svenhjol/Charm#450 opened.

**Comment by YTG1234** [comment]
Fixed in [v0.4.0](https://modrinth.com/mod/recipe-conditions/version/icVsO7lP).

(Note: it's `"frc:conditions"` now)
